### ISSUE TITLE
Chef 13.0.113 - Release April 6, 2017 is not Supported

### ISSUE THREAD

**Issue by Techcadia** [Icinga/chef-icinga2#284]
https://discourse.chef.io/t/chef-client-13-released/10735
```
The Chef::REST API has been removed

It has been fully replaced with Chef::ServerAPI in chef-client code.
```

The error from my run of test kitchen.

```  ================================================================================
       Recipe Compile Error in /tmp/kitchen/cache/cookbooks/icinga2/libraries/search.rb
       ================================================================================

       LoadError
       ---------
       cannot load such file -- chef/rest

       Cookbook Trace:
       ---------------
         /tmp/kitchen/cache/cookbooks/icinga2/libraries/search.rb:23:in `<top (required)>'

       Relevant File Content:
       ----------------------
       /tmp/kitchen/cache/cookbooks/icinga2/libraries/search.rb:

        16:  # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
        17:  # See the License for the specific language governing permissions and
        18:  # limitations under the License.
        19:  #
        20:
        21:  require 'chef'
        22:  require 'chef/node'
        23>> require 'chef/rest'
        24:  require 'chef/role'
        25:  require 'chef/environment'
        26:  require 'chef/data_bag'
        27:  require 'chef/data_bag_item'
        28:  require 'resolv'
        29:
        30:  module Icinga2
        31:    # fetch node information into Hash
        32:    class Search

       System Info:
       ------------
       chef_version=13.0.113
       platform=centos
       platform_version=7.2.1511
       ruby=ruby 2.4.1p111 (2017-03-22 revision 58053) [x86_64-linux]
       program_name=chef-client worker: ppid=11667;start=22:58:58;
       executable=/opt/chef/bin/chef-client
```

Currently I am pinning at the latest 12.x release and the code is functioning correctly.

**Comment by vkhatri** [comment]
Thank you for reporting the issue, will look into it.

**Comment by rnt** [comment]
@vkhatri do you have any news about this?

**Comment by rnt** [comment]
@Techcadia my workarround was add `require_chef_omnibus` to set the chef-client version in `.kitchen.yml`.

```
provisioner:
  name: chef_zero
  always_update_cookbooks: true
  require_chef_omnibus: 12.19.36
```

**Comment by dordal** [comment]
Ran into the same problem today. @vkhatri you think this might be a candidate for an update soon?

**Comment by vills** [comment]
Hehe. Looks like easiest solution - is to remove this line :-).

No code in cookbook use `Chef::REST` class directly (and never used):

```
$: grep -Ri 'chef::rest' . | wc -l
       0

$: git log -p | grep -i 'chef::rest' | wc -l
       0
```

It can be wrong, so better to write tests for libraries before any changes.

**Comment by gsreynolds** [comment]
Unfortunately it looks like Chef 13 compat is a bigger issue than removing the above `require 'chef/rest'` line.

Chef 13 breaks all of the icinga2::server_object_* recipes which uses LWRPs for defining `icinga2_service[generic-service]`, `icinga2_host[generic-host]` etc.

https://github.com/Icinga/chef-icinga2/blob/8fd87273399a68567ffa1115ed4048bc5496936f/libraries/provider_instance.rb#L128-L151

For example for generic-service template under Chef 13, 
```icinga2_objects_dict = {"object"=>{}, "template"=>{}}```
Under Chef 12, 
```icinga2_objects_dict =   {"object"=>{},
 "template"=>
  {"generic-service"=>
    {"import"=>nil,
     "display_name"=>nil,
     "host_name"=>nil,
     "groups"=>nil,
     "check_command"=>nil,
     "max_check_attempts"=>5,
     "check_period"=>nil,
     "check_interval"=>"5m",
     "retry_interval"=>"1m",
     "enable_notifications"=>nil,
     "enable_active_checks"=>nil,
     "enable_passive_checks"=>nil,
     "enable_event_handler"=>nil,
     "enable_flapping"=>true,
     "enable_perfdata"=>nil,
     "event_command"=>nil,
     "flapping_threshold"=>"20",
     "volatile"=>nil,
     "zone"=>nil,
     "command_endpoint"=>nil,
     "notes"=>nil,
     "notes_url"=>nil,
     "action_url"=>nil,
     "icon_image"=>nil,
     "icon_image_alt"=>nil,
     "custom_vars"=>nil,
     "template"=>true,
     "object_class"=>"template"}}}
```

This causes Icinga to fail to restart with: 
```
critical/config: Error: Import references unknown template: 'generic-host'
```
as none of the objects or templates have been written out as the `icinga2_objects_dict['object']` and `icinga2_objects_dict['template']` are empty and only_if guards against this.

Objects dir has none of the expected files defining host/service templates etc.
```ls -lh /etc/icinga2/objects.d/
total 12K
-rw-r----- 1 nagios nagios  295 Aug 15 11:18 envhostgroup_production.conf
-rw-r----- 1 nagios nagios 1.6K Aug 15 11:18 host_production_company.conf
-rw-r----- 1 nagios nagios  287 Aug 15 11:18 idopgsqlconnection.conf```

**Comment by vkhatri** [comment]
@gsreynolds Indeed.

It could be a small change,  perhaps switching to `chef/server_api` api will suffice.

I will do some testing today.

**Comment by vkhatri** [comment]
Fixed in v3.0.0. Please update the issue otherwise.

Closing.
